### PR TITLE
[User Schedule] Implement transform_like and parallelize_like

### PR DESCRIPTION
**Pull request by rdspring1**
This PR implements the `transform_like` and `parallelize_like` functions.

`transform_like(a: Tensor, selected_nodes : [Tensor] = [] )`  propagates the iterDomain transformations to the rest of the fusion. It is a wrapper around `TransformPropagator` and `MaxRootDomainInfoSpanningTree`.

When the `selected_nodes` argument is not empty, it will use the `SetSelector` to apply the transformations to a subset of the graph.

```cpp
TransformPropagator propagator(reference_tv);
SetSelector selector({subgraph_tv_set.begin(), subgraph_tv_set.end()});
MaxRootDomainInfoSpanningTree(reference_tv, &selector).traverse(&propagator);
```

`parallelize_like` is a direct wrapper around `nvfuser::scheduler_utils::parallelizeAllLike`, which will propagate the parallelization scheme to the rest of the fusion.

### PR List:
1. #2282 
2. #2283 
3. #2284
4. #2285  **<<<  YOU ARE HERE**
5. #2286
6. #2287
7. #2288

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
It might be useful to divide `transform_like` into three different functions in a follow-up PR.

`transform_all_like(a: Tensor)` --- Transform all tensors in fusion like reference
`transform_subset_like(a: Tensor, subset_nodes: [Tensor])` --- Transform a subset of tensors like reference
`transform_all_except_like(a: Tensor, exception_subset: [Tensor])` --- Transform all tensor except those in subset like reference.

File: csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -3001,6 +3002,94 @@ void initNvFuserPythonBindings(PyObject* module) {
       },
       py::arg("tensor"),
       py::arg("memory_type"));
+  nvf_sched.def(
+      "transform_like",
+      [](FusionDefinition::SchedOperators& self,
+         Tensor tensor,
+         const std::vector<Tensor>& subgraph_nodes) {
```
Comment (jjsjann123) [inline csrc/python_frontend/python_bindings.cpp:3009]: nitpick: the name of second arg is a bit confusing, it's just `selected_tensors`.

Also, the signature in PR description is wrong.
`transform_like(a: Tensor, subgraph_nodes : None | [Tensor])`. We do NOT have optional on second arg.

Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:3009]: > transform_all_like(a: Tensor) --- Transform all tensors in fusion like reference
transform_subset_like(a: Tensor, subset_nodes: [Tensor]) --- Transform a subset of tensors like reference
transform_all_except_like(a: Tensor, exception_subset: [Tensor]) --- Transform all tensor except those in subset like reference.

Should we just change the signature as
`transform_like(a: Tensor, selected_tensors: [Tensor] = (), excluded_tensors: [Tensor] = ())`. Where `selected_tensors.empty()` would just mean it's all tensors are targeted.

i.e. let's try to keep the APIs somewhat consistent with each other.

File: tests/python/test_schedule_ops.py [hunk 2]
```diff
@@ -315,6 +315,46 @@ def schedule(self):
...
+                cache_after_t0 = fd.sched.cache_after(self.t0)
+                cache_after_t1 = fd.sched.cache_after(self.t1)
+                cache_before_t2 = fd.sched.cache_before(self.t2)
+                fd.sched.set_memory_type(cache_after_t0, MemoryType.shared)
+                fd.sched.set_memory_type(cache_after_t1, MemoryType.shared)
+                fd.sched.set_memory_type(cache_before_t2, MemoryType.shared)
+
+                fd.sched.merge(self.t2, dim=0)
+                fd.sched.split(self.t2, dim=0, factor=128)
+                fd.sched.transform_like(self.t2)
```
Comment (jjsjann123) [inline tests/python/test_schedule_ops.py:347]: if we are going to update the API, let's make sure we actually verify each configurations.

File: csrc/python_frontend/python_bindings.cpp (hunk 1 above)
Comment (rdspring1) [reply csrc/python_frontend/python_bindings.cpp:3009]: I didn't have a need for `transform_all_except_like` with this set of PRs, so I left it alone for now.

**Comment by rdspring1** [comment]
!build

File: tests/python/test_schedule_ops.py (hunk 2 above)
Comment (rdspring1) [reply tests/python/test_schedule_ops.py:459]: I added a test case for partially transformed configuration.

**Comment by rdspring1** [comment]
!build

File: csrc/python_frontend/python_bindings.cpp (hunk 1 above)
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:3000]: more nitpick here. Looks like we updated the description in PR as
`transform_like(a: Tensor, selected_nodes: [Tensor] = [])`

But the signature here is
`transform_like(tensor: Tensor, subgraph_nodes: [Tensor] = [])`

**Review by jjsjann123 — APPROVED** [review]
one minor nitpick on the transform_like signature as well as PR description. Stamping for merge.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Pull request merged**